application/src/main/java/org/thingsboard/server/actors/device/DeviceActorMessageProcessor.java

```diff
@@ -961,19 +963,42 @@ void init(TbActorCtx ctx) {
     }
 
     void checkSessionsTimeout() {
-        log.debug("[{}] checkSessionsTimeout started. Size before check {}", deviceId, sessions.size());
-        long expTime = System.currentTimeMillis() - systemContext.getSessionInactivityTimeout();
-        Map<UUID, SessionInfoMetaData> sessionsToRemove = sessions.entrySet().stream().filter(kv -> kv.getValue().getLastActivityTime() < expTime).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        sessionsToRemove.forEach((sessionId, sessionMD) -> {
-            sessions.remove(sessionId);
-            rpcSubscriptions.remove(sessionId);
-            attributeSubscriptions.remove(sessionId);
-            notifyTransportAboutClosedSession(sessionId, sessionMD, "session timeout!");
-        });
-        if (!sessionsToRemove.isEmpty()) {
-            dumpSessions();
+        final long expTime = System.currentTimeMillis() - systemContext.getSessionInactivityTimeout();
+        List<UUID> expiredIds = null;
+
+        try {
+            for (Map.Entry<UUID, SessionInfoMetaData> kv : sessions.entrySet()) { //entry set are cached for stable sessions
+                if (kv.getValue().getLastActivityTime() < expTime) {
+                    final UUID id = kv.getKey();
+                    if (expiredIds == null) {
+                        expiredIds = new ArrayList<>(1); //most of the expired sessions is a single event
+                    }
+                    expiredIds.add(id);
+
+                }
+            }
+        } catch (ConcurrentModificationException ignored) {
```

Comment:
Please remove this since we are inside an Actor and all messages are processed sequentially.